### PR TITLE
Feature/chapter 5

### PR DESCRIPTION
**Pull request by cbeyer42**

### CODE REVIEW

File: 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    [[ "$output" = *"Gid:"*"("*"0/"*"root"* ]]
+    [[ "$output" = *"Access:"*"(0"[0-7]"00"* ]]
+}
+
+@test "5.3.2 Ensure permissions on SSH private host key files are configured (Automated)" {
+    local SSH_CONFIG=$(find /etc/ssh -xdev -type f -name 'ssh_host_*_key' -exec stat {} \; | grep "Access" | grep "Uid")
+    while IFS= read -r line; do
+        [[ "$line" = *"Uid:"*"("*"0/"*"root"* ]]
+        [[ "$line" = *"Gid:"*"("*"0/"*"root"* ]]
+        [[ "$line" = "Access:"*"(0"[0-7]"00"* ]]
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats]: According to the CIS document permissions have to be "0600 or more restrictive"
```suggestion
        [[ "$line" = "Access:"*"(0"[0-6]"00"* ]]
```

File: 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    run bash -c "sshd -T | grep denygroups"
+    DENYGROUPS=$status
+    [ "$ALLOWUSERS" -eq 0 ] || [ "$ALLOWGROUPS" -eq 0 ] || [ "$DENYUSERS" -eq 0 ] || [ "$DENYGROUPS" -eq 0 ]
+}
+
+@test "5.3.5 Ensure SSH LogLevel is appropriate (Automated)" {
+    run bash -c "sshd -T | grep loglevel"
+    [ "$status" -eq 0 ]
+    [[ "$output" == "LogLevel VERBOSE" ]] || [[ "$output" == "loglevel INFO" ]]
+}
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats:52]: The second check from the CIS document is missing (`grep -is 'loglevel' /etc/ssh/sshd_config /etc/ssh/sshd_config.d/*.conf |
grep -Evi '(VERBOSE|INFO)'`)

File: 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+@test "5.3.6 Ensure SSH X11 forwarding is disabled (Automated)" {
+    run bash -c "sshd -T | grep x11forwarding"
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats]: This command is not according to the CIS document (`sshd -T -C user=root -C host="$(hostname)" -C addr="$(grep $(hostname) /etc/hosts | awk '{print $1}')" | grep -i x11forwarding`)

File: 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    ALLOWGROUPS=$status
+    run bash -c "sshd -T | grep denyusers"
+    DENYUSERS=$status
+    run bash -c "sshd -T | grep denygroups"
+    DENYGROUPS=$status
+    [ "$ALLOWUSERS" -eq 0 ] || [ "$ALLOWGROUPS" -eq 0 ] || [ "$DENYUSERS" -eq 0 ] || [ "$DENYGROUPS" -eq 0 ]
+}
+
+@test "5.3.5 Ensure SSH LogLevel is appropriate (Automated)" {
+    run bash -c "sshd -T | grep loglevel"
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats]: This command is not according to the CIS document (`sshd -T -C user=root -C host="$(hostname)" -C addr="$(grep $(hostname) /etc/hosts | awk '{print $1}')" | grep loglevel`)

**Review by robertauer — CHANGES_REQUESTED** [review]

File: 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    [ "$status" -eq 0 ]
+    [[ "$output" == "LogLevel VERBOSE" ]] || [[ "$output" == "loglevel INFO" ]]
+}
+
+@test "5.3.6 Ensure SSH X11 forwarding is disabled (Automated)" {
+    run bash -c "sshd -T | grep x11forwarding"
+    [ "$status" -eq 0 ]
+    [[ "$output" == "x11forwarding no" ]]
+}
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats:61]: The second check from the CIS document is missing (`grep -Eis '^\s*x11forwarding\s+yes' /etc/ssh/sshd_config
/etc/ssh/sshd_config.d/*.conf`)

File: 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+
+@test "5.3.7 Ensure SSH MaxAuthTries is set to 4 or less (Automated)" {
+    run bash -c "sshd -T | grep maxauthtries"
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats]: This commnand is not according to the CIS document (`sshd -T -C user=root -C host="$(hostname)" -C addr="$(grep $(hostname) /etc/hosts | awk '{print $1}')" | grep maxauthtries`)

File: 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+@test "5.3.11 Ensure SSH PermitEmptyPasswords is disabled (Automated)" {
+    run bash -c "sshd -T | grep permitemptypasswords"
+    [ "$status" -eq 0 ]
+    [[ "$output" == "permitemptypasswords no" ]]
+    run bash -c "grep -Eis '^\s*PermitEmptyPasswords\s+yes' /etc/ssh/sshd_config /etc/ssh/sshd_config.d/*.conf"
+    [[ "$output" == "" ]]
+}
+
+@test "5.3.12 Ensure SSH PermitUserEnvironment is disabled (Automated)" {
+    run bash -c "sshd -T | grep permituserenvironment"
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats]: The first check command in these tests is not according to the CIS document. Is there a reason for that?

File: 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+    [ "$status" -eq 0 ]
+    [[ "$output" == "permituserenvironment no" ]]
+    run bash -c "grep -Eis '^\s*PermitUserEnvironment\s+yes' /etc/ssh/sshd_config /etc/ssh/sshd_config.d/*.conf"
+    [[ "$output" == "" ]]
+}
+
+@test "5.3.13 Ensure only strong Ciphers are used (Automated)" {
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats:117]: Please check this test

File: 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+    [[ "$output" != *"aes256-cbc"* ]]
+    [[ "$output" != *"arcfour"* ]]
+    [[ "$output" != *"arcfour128"* ]]
+    [[ "$output" != *"arcfour256"* ]]
+    [[ "$output" != *"blowfish-cbc"* ]]
+    [[ "$output" != *"cast128-cbc"* ]]
+    [[ "$output" != *"rijndael-cbc@lysator.liu.se"* ]]
+}
+
+@test "5.3.14 Ensure only strong MAC algorithms are used (Automated)" {
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats:126]: Please check this test

File: 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+    [[ "$output" != *"hmac-md5-etm@openssh.com"* ]]
+    [[ "$output" != *"hmac-md5-96-etm@openssh.com"* ]]
+    [[ "$output" != *"hmac-ripemd160-etm@openssh.com"* ]]
+    [[ "$output" != *"hmac-sha1-etm@openssh.com"* ]]
+    [[ "$output" != *"hmac-sha1-96-etm@openssh.com"* ]]
+    [[ "$output" != *"umac-64-etm@openssh.com"* ]]
+    [[ "$output" != *"umac-128-etm@openssh.com"* ]]
+}
+
+@test "5.3.15 Ensure only strong Key Exchange algorithms are used (Automated)" {
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats:135]: Please check this test

File: 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+    run bash -c "sshd -T | grep kexalgorithms"
+    [ "$status" -eq 0 ]
+    [[ "$output" != *"diffie-hellman-group1-sha1"* ]]
+    [[ "$output" != *"diffie-hellman-group14-sha1"* ]]
+    [[ "$output" != *"diffie-hellman-group-exchange-sha1"* ]]
+}
+
+@test "5.3.16 Ensure SSH Idle Timeout Interval is configured (Automated)" {
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats:145]: Please check this test

File: 5-Access-Authentication-and-Authorization/5.4-Configure-PAM.bats [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    PWQUALITY=(${PWQUALITY//minclass = / }) # get the number from the string
+    if [[ "$MINLENGTH" -lt 14 ]] || [[ "$PWQUALITY" -lt 4 ]]; then
+        local CREDITPWQUALITY=$(grep -E '^\s*[duol]credit\s*' /etc/security/pwquality.conf)
+        [[ "$CREDITPWQUALITY" != "" ]]
+        [[ "$CREDITPWQUALITY" == *"dcredit = -1"* ]]
+        [[ "$CREDITPWQUALITY" == *"ucredit = -1"* ]]
+        [[ "$CREDITPWQUALITY" == *"lcredit = -1"* ]]
+        [[ "$CREDITPWQUALITY" == *"ocredit = -1"* ]]
+        local RETRIES=$(grep -E '^\s*password\s+(requisite|required)\s+pam_pwquality\.so\s+(\S+\s+)*retry=[1-3]\s*(\s+\S+\s*)*(\s+#.*)?$' /etc/pam.d/common-password)
+        [[ "$RETRIES" == "password"*"requisite"*"retry="[0-3] ]]
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.4-Configure-PAM.bats:16]: The `[0-3]` part looks like regex to me. If that's true, don't we need `"$RETRIES" =~` here?

File: 5-Access-Authentication-and-Authorization/5.5-User-Accounts-and-Environment.bats [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+load 5.5.4-helper
+load 5.5.5-helper
+
+@test "5.5.1.1 Ensure minimum days between password changes is configured (Automated)" {
+    run bash -c "grep PASS_MIN_DAYS /etc/login.defs | grep --invert-match \"\#\""
+    [ "$status" -eq 0 ]
+    [[ "$output" != "" ]]
+    MINDAYS=(${output//PASS_MIN_DAYS/ }) # get the number from the string
+    [[ "$MINDAYS" -gt 0 ]]
+    run bash -c "grep -E ^[^:]+:[^\!*] /etc/shadow | cut -d: -f1,4"
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.5-User-Accounts-and-Environment.bats]: This command is not according to the CIS document (`awk -F : '(/^[^:]+:[^!*]/ && $4 < 1){print $1 " " $4}' /etc/shadow`)

File: 5-Access-Authentication-and-Authorization/5.5-User-Accounts-and-Environment.bats [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    done <<< "$output"
+}
+
+@test "5.5.1.2 Ensure password expiration is 365 days or less (Automated)" {
+    run bash -c "grep PASS_MAX_DAYS /etc/login.defs | grep --invert-match \"\#\""
+    [ "$status" -eq 0 ]
+    [[ "$output" != "" ]]
+    MAXDAYS=(${output//PASS_MAX_DAYS/ }) # get the number from the string
+    [[ "$MAXDAYS" -lt 366 ]]
+    run bash -c "grep -E '^[^:]+:[^!*]' /etc/shadow | cut -d: -f1,5"
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.5-User-Accounts-and-Environment.bats]: This command is not according to the CIS document (`awk -F: '(/^[^:]+:[^!*]/ && ($5>365 || $5~/([0-1]|-1|\s*)/)){print $1 " "
$5}' /etc/shadow`)

File: 5-Access-Authentication-and-Authorization/5.5-User-Accounts-and-Environment.bats [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+    done <<< "$output"
+}
+
+@test "5.5.1.3 Ensure password expiration warning days is 7 or more (Automated)" {
+    run bash -c "grep PASS_WARN_AGE /etc/login.defs | grep --invert-match \"\#\""
+    [ "$status" -eq 0 ]
+    [[ "$output" != "" ]]
+    WARNAGE=(${output//PASS_WARN_AGE/ }) # get the number from the string
+    [[ "$WARNAGE" -gt 6 ]]
+    run bash -c "grep -E ^[^:]+:[^\!*] /etc/shadow | cut -d: -f1,6"
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.5-User-Accounts-and-Environment.bats]: This command is not according to the CIS document (`awk -F: '(/^[^:]+:[^!*]/ && $6<7){print $1 " " $6}' /etc/shadow`)

File: 5-Access-Authentication-and-Authorization/5.5-User-Accounts-and-Environment.bats [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+@test "5.5.1.4 Ensure inactive password lock is 30 days or less (Automated)" {
+    run bash -c "useradd -D | grep INACTIVE"
+    [ "$status" -eq 0 ]
+    [[ "$output" != "" ]]
+    INACTIVE=(${output//INACTIVE=/ }) # get the number from the string
+    [[ "$INACTIVE" != -1 ]]
+    [[ "$INACTIVE" -lt 31 ]]
+    run bash -c "grep -E ^[^:]+:[^\!*] /etc/shadow | cut -d: -f1,7"
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.5-User-Accounts-and-Environment.bats]: This command is not according to the CIS document (`awk -F: '(/^[^:]+:[^!*]/ && ($7~/(\s*|-1)/ || $7>30)){print $1 " " $7}'
/etc/shadow`)

**Review by robertauer — CHANGES_REQUESTED** [review]
Please check **5.3-Configure-SSH-Server.bats** and **5.5-User-Accounts-and-Environment.bats** completely, there are lots of tests missing or not matching with the CIS document, additional to those I have mentioned here.

File: 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats (hunk 1 above)
Comment (cbeyer42) [reply 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats]: According to the document `group` and `other` should not have any permissions. `600` is the default value and in this case just an example output. So `700` is totally fine in this place.

Comment (robertauer) [reply 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats]: That's what they say in 5.3.1, but in 5.3.2 the document explicitly says `... verify Uid is 0/root and Gid is 0/root and permissions are 0600 or more restrictive`

**Review by robertauer — CHANGES_REQUESTED** [review]

File: 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+    local INTERVAL=$(sshd -T -C user=root -C host="$(hostname)" -C addr="$(grep $(hostname) /etc/hosts | awk '{print $1}')" | grep clientaliveinterval)
+    INTERVAL=(${INTERVAL// / })
+    [[ "${INTERVAL[1]}" -gt 1 ]]
+    [[ "${INTERVAL[1]}" -lt 301 ]]
+    local MAX=$(sshd -T -C user=root -C host="$(hostname)" -C addr="$(grep $(hostname) /etc/hosts | awk '{print $1}')" | grep clientalivecountmax)
+    MAX=(${MAX// / })
+    [[ "${MAX[1]}" -lt 4 ]]
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats:153]: `... verify ClientAliveCountMax is between 1 and 3 :`
```suggestion
    [[ "${MAX[1]}" -gt 0 ]]
    [[ "${MAX[1]}" -lt 4 ]]
```

File: 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+    [[ "$output" == "" ]]
+
+    run bash -c "grep -Ei '^\s*kexalgorithms\s+([^#]+,)?(diffie-hellman-group1-sha1|diffiehellman-group14-sha1|diffie-hellman-group-exchange-sha1)\b' /etc/ssh/sshd_config"
+    [ "$status" -ne 0 ]
+    [[ "$output" == "" ]]
+}
+
+@test "5.3.16 Ensure SSH Idle Timeout Interval is configured (Automated)" {
+    local INTERVAL=$(sshd -T -C user=root -C host="$(hostname)" -C addr="$(grep $(hostname) /etc/hosts | awk '{print $1}')" | grep clientaliveinterval)
+    INTERVAL=(${INTERVAL// / })
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats]: Please briefly describe what is happening in this line and all the similar ones in the following tests.
e.g. in 5.5 there was the `# get the number from the string` comment, which sounds good.
Bash is complicated, help future readers of this :D

File: 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+    run bash -c "grep -Eis '^\s*AllowTcpForwarding\s+yes\b' /etc/ssh/sshd_config /etc/ssh/sshd_config.d/*.conf"
+    [ "$status" -ne 0 ]
+    [[ "$output" == "" ]]
+}
+
+@test "5.3.21 Ensure SSH MaxStartups is configured (Automated)" {
+    run bash -c "sshd -T -C user=root -C host=\"$(hostname)\" -C addr=\"$(grep $(hostname) /etc/hosts | awk '{print $1}')\" | grep -i maxstartups"
+    [ "$status" -eq 0 ]
+    run bash -c "grep -Eis '^\s*maxstartups\s+(((1[1-9]|[1-9][0-9][0-9]+):([0-9]+):([0-9]+))|(([0-9]+):(3[1-9]|[4-9][0-9]|[1-9][0-9][0-9]+):([0-9]+))|(([0-9]+):([0-9]+):(6[1-9]|[7-9][0-9]|[1-9][0-9][0-9]+)))' /etc/ssh/sshd_config /etc/ssh/sshd_config.d/*.conf"
+    [[ "$output" == "" ]]
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats:209]:
```suggestion
    [[ "$output" == "" ]]
    [ "$status" -ne 0 ]
```

File: 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats [hunk 20]
```diff
@@ -0,0 +1,212 @@
...
+    [ "$status" -eq 0 ]
+    [[ "$output" == "allowtcpforwarding no" ]]
+    run bash -c "grep -Eis '^\s*AllowTcpForwarding\s+yes\b' /etc/ssh/sshd_config /etc/ssh/sshd_config.d/*.conf"
+    [ "$status" -ne 0 ]
+    [[ "$output" == "" ]]
+}
+
+@test "5.3.21 Ensure SSH MaxStartups is configured (Automated)" {
+    run bash -c "sshd -T -C user=root -C host=\"$(hostname)\" -C addr=\"$(grep $(hostname) /etc/hosts | awk '{print $1}')\" | grep -i maxstartups"
+    [ "$status" -eq 0 ]
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats:202]: This only checks that the `maxstartups` entry exists, but not that it is `10:30:60 or more restrictive`.

File: 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats [hunk 21]
```diff
@@ -0,0 +1,212 @@
...
+    run bash -c "grep -Eis '^\s*maxstartups\s+(((1[1-9]|[1-9][0-9][0-9]+):([0-9]+):([0-9]+))|(([0-9]+):(3[1-9]|[4-9][0-9]|[1-9][0-9][0-9]+):([0-9]+))|(([0-9]+):([0-9]+):(6[1-9]|[7-9][0-9]|[1-9][0-9][0-9]+)))' /etc/ssh/sshd_config /etc/ssh/sshd_config.d/*.conf"
+    [[ "$output" == "" ]]
+}
+
+@test "5.3.22 Ensure SSH MaxSessions is limited (Automated)" {
+    local MAXSESSIONS=$(sshd -T -C user=root -C host="$(hostname)" -C addr="$(grep $(hostname) /etc/hosts | awk '{print $1}')" | grep -i maxsessions)
+    MAXSESSIONS=(${MAXSESSIONS// / })
+    [[ "${MAXSESSIONS[1]}" -lt 11 ]]
+    run bash -c "grep -Eis '^\s*MaxSessions\s+(1[1-9]|[2-9][0-9]|[1-9][0-9][0-9]+)' /etc/ssh/sshd_config /etc/ssh/sshd_config.d/*.conf"
+    [[ "$output" == "" ]]
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.3-Configure-SSH-Server.bats:218]:
```suggestion
    [[ "$output" == "" ]]
    [ "$status" -ne 0 ]
```

File: 5-Access-Authentication-and-Authorization/5.5-User-Accounts-and-Environment.bats [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+
+@test "5.5.2 Ensure system accounts are secured (Automated)" {
+    run bash -c "awk -F: '\$1!~/(root|sync|shutdown|halt|^\+)/ && \$3<'\"\$(awk '/^\s*UID_MIN/{print \$2}' /etc/login.defs)\"' && \$7!~/((\/usr)?\/sbin\/nologin)/ && \$7!~/(\/bin)?\/false/ {print}' /etc/passwd"
+    [[ "$output" == "" ]]
+    run bash -c "awk -F: '(\$1!~/(root|^\+)/ && \$3<'\"\$(awk '/^\s*UID_MIN/{print \$2}' /etc/login.defs)\"') {print \$1}' /etc/passwd | xargs -I '{}' passwd -S '{}' | awk '(\$2!~/LK?/) {print \$1}'"
+    [[ "$output" == "" ]]
+}
+
+@test "5.5.3 Ensure default group for the root account is GID 0 (Automated)" {
+    run bash -c "grep "^root:" /etc/passwd | cut -f4 -d:"
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.5-User-Accounts-and-Environment.bats]:
```suggestion
    run bash -c "grep \"^root:\" /etc/passwd | cut -f4 -d:"
```

File: 5-Access-Authentication-and-Authorization/5.5-User-Accounts-and-Environment.bats [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+
+@test "5.7 Ensure access to the su command is restricted (Automated)" {
+    run bash -c "grep pam_wheel.so /etc/pam.d/su"
+    [ "$status" -eq 0 ]
+    [[ "$output" == *"auth required pam_wheel.so use_uid group="* ]]
+    filtered_output=$(echo "$output" | grep 'auth required pam_wheel.so use_uid group=')
+    local GROUP=(${filtered_output//auth required pam_wheel.so use_uid group=/ }) # get the group name from the string
+    [[ "$GROUP" != "" ]]
+    run bash -c "grep $GROUP /etc/group"
+    echo "INFO: op -> $output"
```
Comment (robertauer) [inline 5-Access-Authentication-and-Authorization/5.5-User-Accounts-and-Environment.bats]: This may be debug output and can be removed

**Review by robertauer — CHANGES_REQUESTED** [review]

File: 5-Access-Authentication-and-Authorization/5.4-Configure-PAM.bats (hunk 12 above)
Comment (cbeyer42) [reply 5-Access-Authentication-and-Authorization/5.4-Configure-PAM.bats:16]: It works as well with `==`.

Comment (robertauer) [reply 5-Access-Authentication-and-Authorization/5.4-Configure-PAM.bats:16]: But why?

**Review by robertauer — APPROVED** [review]

**Pull request merged**